# Chi-Cuadrado

## <center> $X^2 = \sum_{i=1}^{k}\frac{(Observado_i - esperado_i)^2}{esperado_i}\$

relaciones entre variables cualitativas usando las tablas de contigencias. 
comparar la suma de diferencias al cuadrado entre lo que estamos observando, es decir
nuestros datos reales, y aquello que abria que esperar donde no hay relacion entre variables,
permite cuantificar la magnitud de todas estas distancias y compararlas con la distribucion de chi cuadrado
y decidir si podemos afirmar que estas variables estan relacionadas significativamente.

una de las debilidades de este test es que no permite cuantificar cada una de las relaciones entre categorias, 
pero permite afirmaciones globales como fumar estar relacionado significativamente con el cancer de pulmon, 
y aun así una exploracion visual de las tablas permite determinar en que direccion se da esta relacion.


In [3]:
import pandas as pd
import numpy as np


In [4]:
#load data
df = pd.read_csv('C:/Users/jorge/Desktop/Linkedin DATASCIENCE/CURSO/data/2008.csv')

In [7]:
np.random.seed(0)
df = df[df["Origin"].isin(["HOU","ATL","IND"])]
df = df.sample(frac=1)
df = df[0:10000]

In [8]:
#columna que registra True or False, dependiendo de si el retraso es mayor a x
df["BigDelay"] = df["ArrDelay"] > 30
observados = pd.crosstab(index=df['BigDelay'],columns=df['Origin'], margins = True)

In [9]:
observados

Origin     ATL   HOU  IND    All
BigDelay                        
False     6927   883  765   8575
True      1197   129   99   1425
All       8124  1012  864  10000

6927, es el numero de vuelos que se han retrasado poco y que han salido de Atlanta,
asi mismo para los otros valroes
el margen inf, es la suma por cada una de las columnas
y el margen der, es la suma para cada una de las filas, 
y el total es el total de observaciones. 

In [10]:
from scipy.stats import chi2_contingency

In [11]:
test = chi2_contingency(observados)

In [12]:
test

(8.939538453043031,
 0.17700704816414425,
 6,
 array([[ 6966.33,   867.79,   740.88,  8575.  ],
        [ 1157.67,   144.21,   123.12,  1425.  ],
        [ 8124.  ,  1012.  ,   864.  , 10000.  ]]))

1. valor es el estadistico, la suma de las dif. al cuadrado
2. es el p valor, valor utilizado para tomar decisiones basados en un criterio.
3. tabla de valores esperados

In [13]:
esperados = pd.DataFrame(test[3])

In [14]:
esperados

0        1       2        3
0  6966.33   867.79  740.88   8575.0
1  1157.67   144.21  123.12   1425.0
2  8124.00  1012.00  864.00  10000.0

In [15]:
esperados_rel = round(esperados.apply(lambda r: r/len(df)*100, axis=1),2)

In [16]:
observados_rel = round(observados.apply(lambda r: r/len(df)*100, axis=1),2)

In [17]:
observados_rel

Origin      ATL    HOU   IND     All
BigDelay                            
False     69.27   8.83  7.65   85.75
True      11.97   1.29  0.99   14.25
All       81.24  10.12  8.64  100.00

In [18]:
esperados_rel

0      1     2       3
0  69.66   8.68  7.41   85.75
1  11.58   1.44  1.23   14.25
2  81.24  10.12  8.64  100.00

In [19]:
test[1]

0.17700704816414425

# Resumen de Test de Hipotesis
- si el p-valor < 0.05, hay diferencias significativas: hay relacion entre las variables
- si el p-valor es > 0.05, no hay diferencias significativas: No hay relacion entre las variables

en este caso, es demasiado mayor a 0.05 entonces no podemos afirmar que exista relacion entre variables.

In [20]:
test[1] > 0.05

True